packages/react-router/src/routeRegistry.ts

```diff
@@ -1,156 +1,222 @@
+}
+
+export interface NonIndexRoute extends Omit<NonIndexRouteObject, "children"> {
+    $name?: string;
+    $visibility?: RouteVisibility;
+    children?: Route[];
+}
 
-const IndexToken = "$index$";
+export type Route = IndexRoute | NonIndexRoute;
+
+export type RouteRegistrationStatus = "pending" | "registered";
 
-function normalizePath(routePath: string) {
-    if (routePath !== "/" && routePath.endsWith("/")) {
+function normalizePath(routePath?: string) {
+    if (routePath && routePath !== "/" && routePath.endsWith("/")) {
         return routePath.substring(0, routePath.length - 1);
     }
 
     return routePath;
 }
 
-export function createIndexKey(route: Route, layoutPath: string) {
-    if (route.index) {
-        return layoutPath.endsWith("/") ? `${layoutPath}${IndexToken}` : `${layoutPath}/${IndexToken}`;
+export function createIndexKey(route: Route) {
+    if (route.path) {
+        return normalizePath(route.path);
     }
 
-    return normalizePath(route.path!);
-}
-
-export type RouteRegistrationStatus = "pending" | "registered";
+    if (route.$name) {
+        return route.$name;
+    }
 
-export interface AddRouteReturnType {
-    registrationStatus: RouteRegistrationStatus;
-    completedPendingRegistrations?: Route[];
+    return undefined;
 }
 
+
 export class RouteRegistry {
-    #routes: RootRoute[];
+    #routes: Route[];
 
     // Using an index to speed up the look up of parent routes.
-    // <indexKey, RootRoute | Route>
-    readonly #routesIndex: Map<string, RootRoute | Route> = new Map();
+    // <indexKey, Route>
+    readonly #routesIndex: Map<string, Route> = new Map();
 
     // A collection of pending routes to registered once their layout is registered.
-    // <layoutPath, Route[]>
+    // <parentPath | parentName, Route[]>
     readonly #pendingRegistrations: Map<string, Route[]> = new Map();
 
     constructor() {
         this.#routes = [];
     }
 
-    add(routes: RootRoute[] | Route[], { layoutPath }: RegisterRoutesOptions = {}) {
-        if (layoutPath) {
-            return this.#addNestedRoutes(routes, layoutPath);
+    #addIndex(route: Route) {
+        const key = createIndexKey(route);
+
+        if (key) {
+            if (this.#routesIndex.has(key)) {
+                throw new Error(`[squide] A route index has already been registered for the key: "${key}". Did you register two routes with the same "path" or "name" property?`);
+            }
+
+            this.#routesIndex.set(key, route);
         }
 
-        return this.#addRootRoutes(routes);
+        return key;
     }
 
-    #addRootRoutes(routes: RootRoute[]): AddRouteReturnType {
-        // Add index entries to speed up the registration of future nested routes.
-        routes.forEach(x => {
-            const key = createIndexKey(x, "/");
+    #recursivelyAddRoutes(routes: Route[]) {
+        const newRoutes: Route[] = [];
+        const completedPendingRegistrations: Route[] = [];
+
+        routes.forEach((x: Route) => {
```

Comment:
a forEach on an array that creates a new array with the same amount of items should probably converted to a map

Reply:
I dont know, that was also my initial thought but this loop is doing more than mapping route that's why I preferer a ``forEach`.

On top of mapping the route, this loop is also:
- Creating indexes for the encountered routes
- Registering pending routes that were waiting for this route to be registered

That kind of feel weird to do all this in a `map` as it is a lot of side effects.

Reply:
Poor old [for..of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of), unappreciated and unloved, when it could do the job perfectly fine 😂

Reply:
Hehe, how different from `forEach` would it be? I usually go for `for..of` when I have to break the loop given that a condition is met.